# <center>Advanced Searching on Strings</center>

<center>Dr. W.J.B. Mattingly</center>

<center>Smithsonian Data Science Lab and United States Holocaust Memorial Museum</center>

<center>August 2021</center>

## Covered in this Chapter

1) How to find Strings with Specific Features<br>
2) Finding Strings without Certain Features<br>
3) Larger Pandas Queries with RegEx<br>

## Finding Features within a String

In [1]:
import pandas as pd
df = pd.read_csv("data/titanic.csv")
df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

When I am looking at the df, I notice that there is a "Rev." in index 886. As a historian, I find this fascinating. Now, I start to wonder, how many reverends were there on the Titanic? Is this individual unique? If I wanted to ask this question outside of Pandas, I could do the following:

In [2]:
names = df.Name.tolist()
revs = []
for name in names:
    if "Rev." in name:
        revs.append(name)
print (revs)

['Byles, Rev. Thomas Roussel Davids', 'Bateman, Rev. Robert James', 'Carter, Rev. Ernest Courtenay', 'Kirkland, Rev. Charles Leonard', 'Harper, Rev. John', 'Montvila, Rev. Juozas']


Sure, that works, but I don't have any of the other data associated with each of these reverends. I would have to then do some manual searching in the DataFrame to find their corresponding data, or save the data as a dictionary and then run look ups. But why do all of that, when we can do it in a single line of code using Pandas' built-in function. We can use  .str.contains() which takes an argument of what we want to return.

In [3]:
df.loc[df["Name"].str.contains("Rev.")]

PassengerId  Survived  Pclass                               Name   Sex  \
149          150         0       2  Byles, Rev. Thomas Roussel Davids  male   
150          151         0       2         Bateman, Rev. Robert James  male   
249          250         0       2      Carter, Rev. Ernest Courtenay  male   
626          627         0       2     Kirkland, Rev. Charles Leonard  male   
848          849         0       2                  Harper, Rev. John  male   
886          887         0       2              Montvila, Rev. Juozas  male   

      Age  SibSp  Parch       Ticket    Fare Cabin Embarked  
149  42.0      0      0       244310  13.000   NaN        S  
150  51.0      0      0  S.O.P. 1166  12.525   NaN        S  
249  54.0      1      0       244252  26.000   NaN        S  
626  57.0      0      0       219533  12.350   NaN        Q  
848  28.0      0      1       248727  33.000   NaN        S  
886  27.0      0      0       211536  13.000   NaN        S

We can, therefore, see not only the reverends, but also their corresponding data.

## Finding Strings that Don't Contain Feature

What if we wanted to eliminate all names that do not contain "Rev."? We can introduce "~" prior to df to specify that the Names column should not have whatever condition we express.

In [4]:
df.loc[~df["Name"].str.contains("Rev.")]

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
885          886         0       3   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
885               Rice, Mrs. William (Margaret Norton)  female  39.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
885      5            382652  29.1250   NaN        Q  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[885 rows x 12 columns]

## Using RegEx with Pandas

Out of the box, Pandas supports RegEx. RegEx stands for Regular Expressions. It is a powerful way of performing complex string matching. If we were interested in finding any instance of "Rev." or "Mr.", we would have to write something like this without RegEx:

In [5]:
df.loc[(df["Name"].str.contains("Rev.")) | (df["Name"].str.contains("Mr."))]

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
3              4         1       1   
4              5         0       3   
5              6         0       3   
..           ...       ...     ...   
884          885         0       3   
885          886         0       3   
886          887         0       2   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
5                                     Moran, Mr. James    male   NaN      0   
..                                                 ...     ...   ...    ...   
884                             Sutehall, Mr. Henry Jr    male  25.0      0   
885               Rice, Mrs. William (Margaret Norton)  female  39.0      0   
886                              Montvila, Rev. Juozas    male  27.0      0   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch           Ticket     Fare Cabin Embarked  
0        0        A/5 21171   7.2500   NaN        S  
1        0         PC 17599  71.2833   C85        C  
3        0           113803  53.1000  C123        S  
4        0           373450   8.0500   NaN        S  
5        0           330877   8.4583   NaN        Q  
..     ...              ...      ...   ...      ...  
884      0  SOTON/OQ 392076   7.0500   NaN        S  
885      5           382652  29.1250   NaN        Q  
886      0           211536  13.0000   NaN        S  
889      0           111369  30.0000  C148        C  
890      0           370376   7.7500   NaN        Q  

[653 rows x 12 columns]

While this works, imagine if we had 20 or 30 different conditions! That would be a very long piece of code to write and while it would work, it is always best practice to write shorter, tighter code. So, let's do the same thing, but with RegEx. We can add the Or-condition into the str.contains() argument. This is a RegEx command. To ensure that RegEx is registered, it may be necessary to pass it as an argument.

In [6]:
df.loc[df["Name"].str.contains("Rev.|Mr.", regex=True)]

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
3              4         1       1   
4              5         0       3   
5              6         0       3   
..           ...       ...     ...   
884          885         0       3   
885          886         0       3   
886          887         0       2   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
5                                     Moran, Mr. James    male   NaN      0   
..                                                 ...     ...   ...    ...   
884                             Sutehall, Mr. Henry Jr    male  25.0      0   
885               Rice, Mrs. William (Margaret Norton)  female  39.0      0   
886                              Montvila, Rev. Juozas    male  27.0      0   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch           Ticket     Fare Cabin Embarked  
0        0        A/5 21171   7.2500   NaN        S  
1        0         PC 17599  71.2833   C85        C  
3        0           113803  53.1000  C123        S  
4        0           373450   8.0500   NaN        S  
5        0           330877   8.4583   NaN        Q  
..     ...              ...      ...   ...      ...  
884      0  SOTON/OQ 392076   7.0500   NaN        S  
885      5           382652  29.1250   NaN        Q  
886      0           211536  13.0000   NaN        S  
889      0           111369  30.0000  C148        C  
890      0           370376   7.7500   NaN        Q  

[653 rows x 12 columns]

In some instances, we may have uncleaned data and the use of "Rev." may be lowercase in one instance. To ensure that we grab both upper and lowercase forms of this sequence, let's ignore the case by using the flags keyword argument.

In [7]:
import re
df.loc[df["Name"].str.contains("Rev.|Mr.", flags=re.I, regex=True)]

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
3              4         1       1   
4              5         0       3   
5              6         0       3   
..           ...       ...     ...   
884          885         0       3   
885          886         0       3   
886          887         0       2   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
5                                     Moran, Mr. James    male   NaN      0   
..                                                 ...     ...   ...    ...   
884                             Sutehall, Mr. Henry Jr    male  25.0      0   
885               Rice, Mrs. William (Margaret Norton)  female  39.0      0   
886                              Montvila, Rev. Juozas    male  27.0      0   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch           Ticket     Fare Cabin Embarked  
0        0        A/5 21171   7.2500   NaN        S  
1        0         PC 17599  71.2833   C85        C  
3        0           113803  53.1000  C123        S  
4        0           373450   8.0500   NaN        S  
5        0           330877   8.4583   NaN        Q  
..     ...              ...      ...   ...      ...  
884      0  SOTON/OQ 392076   7.0500   NaN        S  
885      5           382652  29.1250   NaN        Q  
886      0           211536  13.0000   NaN        S  
889      0           111369  30.0000  C148        C  
890      0           370376   7.7500   NaN        Q  

[654 rows x 12 columns]